doc/smp_write_gpio.8
.TH SMP_WRITE_GPIO "8" "August 2011" "smp_utils\-0.97" SMP_UTILS
.SH NAME
smp_write_gpio \- invoke WRITE GPIO REGISTER (ENHANCED) SMP function
.SH SYNOPSIS
.B smp_write_gpio
[\fI\-\-count=CO\fR] [\fI\-\-data=H,H...\fR] [\fI\-\-help\fR] [\fI\-\-hex\fR]
[\fI\-\-index=IN\fR] [\fI\-\-interface=PARAMS\fR] [\fI\-\-raw\fR]
[\fI\-\-sa=SAS_ADDR\fR] [\fI\-\-type=TY\fR] [\fI\-\-verbose\fR]
[\fI\-\-version\fR] \fISMP_DEVICE[,N]\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
Sends a SAS Serial Management Protocol (SMP) WRITE GPIO REGISTER or a WRITE
GPIO REGISTER ENHANCED function request to an SMP target. The SMP target is
identified by the \fISMP_DEVICE\fR and the \fISAS_ADDR\fR. Depending on the
interface, the \fISAS_ADDR\fR may be deduced from the \fISMP_DEVICE\fR. The
mpt interface uses \fISMP_DEVICE\fR to identify a HBA (an SMP initiator)
and needs the additional \fI,N\fR to differentiate between HBAs if there
are multiple present.
.PP
This function is defined in SFF\-8485 ( see www.sffcommittee.com ) together
with its corresponding READ GPIO REGISTER (ENHANCED) SMP function.
The other SMP functions are defined in SAS documents at www.t10.org .
SFF\-8485 defines a curious device called a "virtual SMP port"
that lives in a host bus adapter (HBA) and allows SMP to manipulate
the sideband signals in wide internal cables. To stop other initiators
in a multi\-initiator domain from accessing those sideband signals,
the virtual SMP (target) port is not indicated in a DISCOVER response.
.PP
For notes on the SMP WRITE GPIO REGISTER ENHANCED function see the section 
on the ENHANCED FUNCTION below.
.SH OPTIONS
Mandatory arguments to long options are mandatory for short options as well.
.TP
\fB\-c\fR, \fB\-\-count\fR=\fICO\fR
where \fICO\fR is the register count. This is the number of (4 byte)
registers to send. The default value is 1.
.TP
\fB\-d\fR, \fB\-\-data\fR=\fIH,H...\fR
this option supplies a comma separated list of hexadecimal bytes (0 to
ff inclusive) that will be used as write data in the WRITE GPIO REGISTER
request. The number of bytes supplied should be four times \fI\-\-count=C\fR
The first 4 bytes will become the first register written. 
.TP
\fB\-d\fR, \fB\-\-data\fR=\fB-\fR
reads string of hexadecimal bytes from stdin. Spaces, tabs and line feeds
additionally are permitted as separators.
.TP
\fB\-E\fR, \fB\-\-enhanced\fR
sends a WRITE GPIO REGISTER ENHANCED function request (default: send a WRITE
GPIO REGISTER function request).
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-H\fR, \fB\-\-hex\fR
output the response (less the CRC field) in hexadecimal.
.TP
\fB\-i\fR, \fB\-\-index\fR=\fIIN\fR
where \fIIN\fR is the register index. This value is origin zero and its
default value is 0. The request will send \fI\-\-count=CO\fR registers
starting from \fI\-\-index=IN\fR of the given \fI\-\-type=TY\fR.
.TP
\fB\-I\fR, \fB\-\-interface\fR=\fIPARAMS\fR
interface specific parameters. In this case "interface" refers to the
path through the operating system to the SMP initiator. See the smp_utils
man page for more information.
.TP
\fB\-r\fR, \fB\-\-raw\fR
send the response (less the CRC field) to stdout in binary. All error
messages are sent to stderr.
.TP
\fB\-s\fR, \fB\-\-sa\fR=\fISAS_ADDR\fR
specifies the SAS address of the SMP target device. Typically this is an
expander. This option may not be needed if the \fISMP_DEVICE\fR has the
target's SAS address within it. The \fISAS_ADDR\fR is in decimal but most
SAS addresses are shown in hexadecimal. To give a number in hexadecimal
either prefix it with '0x' or put a trailing 'h' on it.
.TP
\fB\-t\fR, \fB\-\-type\fR=\fITY\fR
where \fITY\fR is the register type. The default value is 0 (GPIO_CFG).
The request will send \fI\-\-count=CO\fR registers starting from
\fI\-\-index=IN\fR of the given \fI\-\-type=TY\fR.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the verbosity of the output. Can be used multiple times.
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH ENHANCED FUNCTION
In the technical review of SAS\-2 prior to standardization in this t10
document: 08\-212r8.pdf (page 871 or 552) there is a comment that the
READ GPIO REGISTER and WRITE GPIO REGISTER function headers (i.e. first
4 bytes) do not comply with all the other SMP functions in SAS\-2.
.PP
There is a suggestion that enhanced variants be introduced in the next
version of the SFF\-8485 document. If that ever happened then it has
not been made public. In SAS\-2.1 both the READ and WRITE GPIO REGISTER
functions have been made obsolete and the corresponding ENHANCED function
numbers are "restricted" for SFF\-8485.
.PP
If the \fI\-\-enhanced\fR option is given then the ENHANCED function 
number is sent; the register type, index and count fields (1  byte each) 
are increased by two byte positions leaving enough space for a compliant
SAS\-2 SMP header to be built.
.SH CONFORMING TO
The SMP WRITE GPIO REGISTER function is defined in SFF\-8485 (see
www.sffcommittee.com ) together with its corresponding READ
GPIO REGISTER SMP function.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2006\-2011 Douglas Gilbert
.br
This software is distributed under a FreeBSD license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B smp_utils, smp_read_gpio(smp_utils)


NEWS
See ChangeLog, README and INSTALL files.
There is also a html page at http://sg.danny.cz/sg/smp_utils.html
and an overview man page in this file: doc/smp_utils.8


src/smp_write_gpio.c
/*
 * Copyright (c) 2006-2011 Douglas Gilbert.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include "smp_lib.h"

/* This is a Serial Attached SCSI (SAS) Serial Management Protocol (SMP)
 * utility.
 *
 * This utility issues a WRITE GPIO REGISTER request and outputs its
 * response.  The WRITE GPIO REGISTER function is defined in SFF-8485
 * 0.7 [2006/02/01].
 *
 * Based on a comment by Rob Elliott in this t10 document : 08-212r8.pdf
 * (page 871 or 552) the ENHANCED variant request has its 4-byte header
 * changed to comply with other SAS-2 SMP requests. This will increase
 * the byte position by 2 of the register type, index and count fields.
 * The remaining write data (first..last register) is not moved.
 */

static char * version_str = "1.09 20111222";

#define SMP_MAX_REQ_LEN (1020 + 4 + 4)

static struct option long_options[] = {
    {"count", 1, 0, 'c'},
    {"data", 1, 0, 'd'},
    {"enhanced", 0, 0, 'E'},
    {"help", 0, 0, 'h'},
    {"hex", 0, 0, 'H'},
    {"index", 1, 0, 'i'},
    {"interface", 1, 0, 'I'},
    {"phy", 1, 0, 'p'},
    {"raw", 0, 0, 'r'},
    {"sa", 1, 0, 's'},
    {"type", 0, 0, 't'},
    {"verbose", 0, 0, 'v'},
    {"version", 0, 0, 'V'},
    {0, 0, 0, 0},
};


static void
usage(void)
{
    fprintf(stderr, "Usage: "
          "smp_write_gpio [--count=CO] [--data=H,H...] [--enhanced] "
          "[--help]\n"
          "                      [--hex] [--index=IN] [--interface=PARAMS] "
          "[--raw]\n"
          "                      [--sa=SAS_ADDR] [type=TY] [--verbose] "
          "[--version]\n"
          "                      SMP_DEVICE[,N]\n"
          "  where:\n"
          "    --count=CO|-c CO     register count (dwords to write) "
          "(def: 1)\n"
          "    --data=H,H...|-d H,H...    comma separated list of hex "
          "bytes to write\n"
          "    --data=-|-d -        read stdin for hex bytes to write\n"
          "    --enhanced|-E        use WRITE GPIO REGISTER ENHANCED "
          "function\n"
          "    --help|-h            print out usage message\n"
          "    --hex|-H             print response in hexadecimal\n"
          "    --index=IN|-i IN     register index (def: 0)\n"
          "    --interface=PARAMS|-I PARAMS    specify or override "
          "interface\n"
          "    --raw|-r             output response in binary\n"
          "    --sa=SAS_ADDR|-s SAS_ADDR    SAS address of SMP "
          "target (use leading\n"
          "                                 '0x' or trailing 'h'). "
          "Depending on\n"
          "                                 the interface, may not be "
          "needed\n"
          "    --type=TY|-t TY      register type (def: 0 (GPIO_CFG))\n"
          "    --verbose|-v         increase verbosity\n"
          "    --version|-V         print version string and exit\n\n"
          "Performs a SMP WRITE GPIO REGISTER (default) or SMP WRITE GPIO "
          "REGISTER\nENHANCED function\n"
          );
}

static void
dStrRaw(const char* str, int len)
{
    int k;

    for (k = 0 ; k < len; ++k)
        printf("%c", str[k]);
}

static int
read_hex(const char * inp, unsigned char * arr, int * arr_len)
{
    int in_len, k, j, m, off;
    unsigned int h;
    const char * lcp;
    char * cp;
    char line[1024];

    if ((NULL == inp) || (NULL == arr) || (NULL == arr_len))
        return 1;
    lcp = inp;
    in_len = strlen(inp);
    if (0 == in_len) {
        *arr_len = 0;
    }
    if ('-' == inp[0]) {        /* read from stdin */
        for (j = 0, off = 0; j < 1024; ++j) {
            if (NULL == fgets(line, sizeof(line), stdin))
                break;
            in_len = strlen(line);
            if (in_len > 0) {
                if ('\n' == line[in_len - 1]) {
                    --in_len;
                    line[in_len] = '\0';
                }
            }
            if (0 == in_len)
                continue;
            lcp = line;
            m = strspn(lcp, " \t");
            if (m == in_len)
                continue;
            lcp += m;
            in_len -= m;
            if ('#' == *lcp)
                continue;
            k = strspn(lcp, "0123456789aAbBcCdDeEfF ,\t");
            if (in_len != k) {
                fprintf(stderr, "read_hex: syntax error at "
                        "line %d, pos %d\n", j + 1, m + k + 1);
                return 1;
            }
            for (k = 0; k < 1024; ++k) {
                if (1 == sscanf(lcp, "%x", &h)) {
                    if (h > 0xff) {
                        fprintf(stderr, "read_hex: hex number "
                                "larger than 0xff in line %d, pos %d\n",
                                j + 1, (int)(lcp - line + 1));
                        return 1;
                    }
                    arr[off + k] = h;
                    lcp = strpbrk(lcp, " ,\t");
                    if (NULL == lcp)
                        break;
                    lcp += strspn(lcp, " ,\t");
                    if ('\0' == *lcp)
                        break;
                } else {
                    fprintf(stderr, "read_hex: error in "
                            "line %d, at pos %d\n", j + 1,
                            (int)(lcp - line + 1));
                    return 1;
                }
            }
            off += k + 1;
        }
        *arr_len = off;
    } else {        /* hex string on command line */
        k = strspn(inp, "0123456789aAbBcCdDeEfF,");
        if (in_len != k) {
            fprintf(stderr, "read_hex: error at pos %d\n",
                    k + 1);
            return 1;
        }
        for (k = 0; k < 1024; ++k) {
            if (1 == sscanf(lcp, "%x", &h)) {
                if (h > 0xff) {
                    fprintf(stderr, "read_hex: hex number larger "
                            "than 0xff at pos %d\n", (int)(lcp - inp + 1));
                    return 1;
                }
                arr[k] = h;
                cp = strchr(lcp, ',');
                if (NULL == cp)
                    break;
                lcp = cp + 1;
            } else {
                fprintf(stderr, "read_hex: error at pos %d\n",
                        (int)(lcp - inp + 1));
                return 1;
            }
        }
        *arr_len = k + 1;
    }
    return 0;
}


int
main(int argc, char * argv[])
{
    int res, c, k, len, act_resplen, off;
    int rcount = 1;
    int do_data = 0;
    int enhanced = 0;
    int do_hex = 0;
    int rindex = 0;
    int phy_id = 0;
    int do_raw = 0;
    int rtype = 0;
    int verbose = 0;
    long long sa_ll;
    unsigned long long sa = 0;
    char i_params[256];
    char device_name[512];
    unsigned char smp_req[SMP_MAX_REQ_LEN];
    unsigned char smp_resp[8];
    struct smp_target_obj tobj;
    struct smp_req_resp smp_rr;
    unsigned char data_arr[1024];
    int arr_len = 0;
    int subvalue = 0;
    char * cp;
    char b[128];
    int ret = 0;

    memset(device_name, 0, sizeof device_name);
    memset(i_params, 0, sizeof i_params);
    memset(smp_req, 0, sizeof smp_req);
    smp_req[0] = SMP_FRAME_TYPE_REQ;
    smp_req[1] = SMP_FN_WRITE_GPIO_REG;
    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "c:d:EhHi:I:p:rs:t:vV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'c':
            rcount = smp_get_num(optarg);
            if ((rcount < 1) || (rcount > 255)) {
                fprintf(stderr, "bad argument to '--count'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'd':
            memset(data_arr, 0, sizeof(data_arr));
            if (read_hex(optarg, data_arr, &arr_len)) {
                fprintf(stderr, "bad argument to '--data'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            do_data = 1;
            break;
        case 'E':
            ++enhanced;
            break;
        case 'h':
        case '?':
            usage();
            return 0;
        case 'H':
            ++do_hex;
            break;
        case 'i':
            rindex = smp_get_num(optarg);
            if ((rindex < 0) || (rindex > 255)) {
                fprintf(stderr, "bad argument to '--index'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'I':
            strncpy(i_params, optarg, sizeof(i_params));
            i_params[sizeof(i_params) - 1] = '\0';
            break;
        case 'p':
           phy_id = smp_get_num(optarg);
           if ((phy_id < 0) || (phy_id > 254)) {
                fprintf(stderr, "bad argument to '--phy', expect "
                        "value from 0 to 254\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            if (verbose)
                fprintf(stderr, "'--phy=<n>' option not needed so "
                        "ignored\n");
            break;
        case 'r':
            ++do_raw;
            break;
        case 's':
           sa_ll = smp_get_llnum(optarg);
           if (-1LL == sa_ll) {
                fprintf(stderr, "bad argument to '--sa'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            sa = (unsigned long long)sa_ll;
            break;
        case 't':
            rtype = smp_get_num(optarg);
            if ((rtype < 0) || (rtype > 255)) {
                fprintf(stderr, "bad argument to '--type'\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
            break;
        case 'v':
            ++verbose;
            break;
        case 'V':
            fprintf(stderr, "version: %s\n", version_str);
            return 0;
        default:
            fprintf(stderr, "unrecognised switch code 0x%x ??\n", c);
            usage();
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if ('\0' == device_name[0]) {
            strncpy(device_name, argv[optind], sizeof(device_name) - 1);
            device_name[sizeof(device_name) - 1] = '\0';
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                fprintf(stderr, "Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if ((0 == do_data) || (arr_len < 1)) {
        fprintf(stderr, "need to supply data to write, see "
                "'--data=' option\n");
        usage();
        return SMP_LIB_SYNTAX_ERROR;
    }
    if ((rcount * 4) != arr_len) {
        fprintf(stderr, "number of data bytes given (%d) needs "
                "to be 4 times count (%d)\n", arr_len, rcount);
        return SMP_LIB_SYNTAX_ERROR;
    }
    if (0 == device_name[0]) {
        cp = getenv("SMP_UTILS_DEVICE");
        if (cp)
            strncpy(device_name, cp, sizeof(device_name) - 1);
        else {
            fprintf(stderr, "missing device name on command line\n    [Could "
                    "use environment variable SMP_UTILS_DEVICE instead]\n");
            usage();
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if ((cp = strchr(device_name, SMP_SUBVALUE_SEPARATOR))) {
        *cp = '\0';
        if (1 != sscanf(cp + 1, "%d", &subvalue)) {
            fprintf(stderr, "expected number after separator in SMP_DEVICE "
                    "name\n");
            return SMP_LIB_SYNTAX_ERROR;
        }
    }
    if (0 == sa) {
        cp = getenv("SMP_UTILS_SAS_ADDR");
        if (cp) {
           sa_ll = smp_get_llnum(cp);
           if (-1LL == sa_ll) {
                fprintf(stderr, "bad value in environment variable "
                        "SMP_UTILS_SAS_ADDR\n");
                fprintf(stderr, "    use 0\n");
                sa_ll = 0;
            }
            sa = (unsigned long long)sa_ll;
        }
    }
    if (sa > 0) {
        if (! smp_is_naa5(sa)) {
            fprintf(stderr, "SAS (target) address not in naa-5 format "
                    "(may need leading '0x')\n");
            if ('\0' == i_params[0]) {
                fprintf(stderr, "    use '--interface=' to override\n");
                return SMP_LIB_SYNTAX_ERROR;
            }
        }
    }

    res = smp_initiator_open(device_name, subvalue, i_params, sa,
                             &tobj, verbose);
    if (res < 0)
        return SMP_LIB_FILE_ERROR;
    if (enhanced) {
        smp_req[1] = SMP_FN_WRITE_GPIO_REG_ENH;
        smp_req[2] = 0x0;       /* response is only header+CRC */
        smp_req[3] = rcount + 1;/* request: extra dword for register fields */
        off = 2;
    } else
        off = 0;
    smp_req[2 + off] = rtype;
    smp_req[3 + off] = rindex;
    smp_req[4 + off] = rcount;
    for (k = 0; k < arr_len; ++k)
        smp_req[8 + k] = data_arr[k];
    if (verbose) {
        fprintf(stderr, "    Write GPIO register%s request: ",
                (enhanced ? " enhanced" : ""));
        for (k = 0; k < (12 + arr_len); ++k)
            fprintf(stderr, "%02x ", smp_req[k]);
        fprintf(stderr, "\n");
    }
    memset(&smp_rr, 0, sizeof(smp_rr));
    smp_rr.request_len = 12 + arr_len;
    smp_rr.request = smp_req;
    smp_rr.max_response_len = sizeof(smp_resp);
    smp_rr.response = smp_resp;
    res = smp_send_req(&tobj, &smp_rr, verbose);

    if (res) {
        fprintf(stderr, "smp_send_req failed, res=%d\n", res);
        if (0 == verbose)
            fprintf(stderr, "    try adding '-v' option for more debug\n");
        ret = -1;
        goto err_out;
    }
    if (smp_rr.transport_err) {
        fprintf(stderr, "smp_send_req transport_error=%d\n",
                smp_rr.transport_err);
        ret = -1;
        goto err_out;
    }
    act_resplen = smp_rr.act_response_len;
    if ((act_resplen >= 0) && (act_resplen < 4)) {
        fprintf(stderr, "response too short, len=%d\n", act_resplen);
        ret = SMP_LIB_CAT_MALFORMED;
        goto err_out;
    }
    len = 4 + (rcount * 4);      /* length in bytes, excluding 4 byte CRC */
    if ((act_resplen >= 0) && (len > act_resplen)) {
        if (verbose)
            fprintf(stderr, "actual response length [%d] less than deduced "
                    "length [%d]\n", act_resplen, len);
        len = act_resplen;
    }
    if (do_hex || do_raw) {
        if (do_hex)
            dStrHex((const char *)smp_resp, len, 1);
        else
            dStrRaw((const char *)smp_resp, len);
        if (SMP_FRAME_TYPE_RESP != smp_resp[0])
            ret = SMP_LIB_CAT_MALFORMED;
        else if (smp_resp[1] != smp_req[1])
            ret = SMP_LIB_CAT_MALFORMED;
        else if (smp_resp[2])
            ret = smp_resp[2];
        goto err_out;
    }
    if (SMP_FRAME_TYPE_RESP != smp_resp[0]) {
        fprintf(stderr, "expected SMP frame response type, got=0x%x\n",
                smp_resp[0]);
        ret = SMP_LIB_CAT_MALFORMED;
        goto err_out;
    }
    if (smp_resp[1] != smp_req[1]) {
        fprintf(stderr, "Expected function code=0x%x, got=0x%x\n",
                smp_req[1], smp_resp[1]);
        ret = SMP_LIB_CAT_MALFORMED;
        goto err_out;
    }
    if (smp_resp[2]) {
        ret = smp_resp[2];
        cp = smp_get_func_res_str(ret, sizeof(b), b);
        fprintf(stderr, "Write gpio register%s result: %s\n",
                (enhanced ? " enhanced" : ""), cp);
        goto err_out;
    }

err_out:
    res = smp_initiator_close(&tobj);
    if (res < 0) {
        if (0 == ret)
            return SMP_LIB_FILE_ERROR;
    }
    if (ret < 0)
        ret = SMP_LIB_CAT_OTHER;
    if (verbose && ret)
        fprintf(stderr, "Exit status %d indicates error detected\n", ret);
    return ret;
}
